src/flyte/app/_input.py

```diff
@@ -1,19 +1,162 @@
 from __future__ import annotations
 
+import os
 import re
 import typing
 from dataclasses import dataclass, field
 from functools import cache, cached_property
 from typing import List, Literal, Optional
 
-from pydantic import BaseModel
+from pydantic import BaseModel, model_validator
 
-if typing.TYPE_CHECKING:
-    import flyte.io
+import flyte.io
+from flyte._initialize import requires_initialization
+from flyte.remote._task import AutoVersioning
+
+InputTypes = str | flyte.io.File | flyte.io.Dir
+_SerializedInputType = Literal["string", "file", "directory"]
+
+INPUT_TYPE_MAP = {
+    str: "string",
+    flyte.io.File: "file",
+    flyte.io.Dir: "directory",
+}
 
 RUNTIME_INPUTS_FILE = "flyte-inputs.json"
 
-InputType = Literal["file", "directory", "string"]
+
+class _DelayedValue(BaseModel):
+    """
+    Delayed value for app inputs.
+    """
+
+    type: _SerializedInputType
+
+    @model_validator(mode="before")
+    @classmethod
+    def check_type(cls, data: typing.Any) -> typing.Any:
+        if "type" in data:
+            data["type"] = INPUT_TYPE_MAP.get(data["type"], data["type"])
+        return data
+
+    async def get(self) -> str:
```

Comment:
Can we put some logs, in debug mode and also a warning log, after it is materialized for the name of the key that was materialized from remote.